docs/proposals/14-adminkubeconfig-subresource.md

```diff
@@ -0,0 +1,103 @@
+
+### Goals
+
+- Add a `Shoot` subresource called `adminkubeconfig` that would produce a `kubeconfig` used to access that `Shoot` cluster.
+- The `kubeconfig` is not stored in the API Server, but generated for each request.
+- In the `AdminKubeConfigRequest` send to that subresource, end-users can specify the expiration time of the credential.
+- The identity (user) in the Gardener cluster would be part of the identity (x509 client certificate). E.g if `Joe` authenticates against the Gardener API server, the generated certificate for `Shoot` authentication would have the following subject:
+
+  - Common Name: `Joe`
+  - Organisation: `system:masters`
+
+- The maximum validity of the certificate can be enforced by setting a flag on the `gardener-apiserver`.
+- Deprecate and remove the old `{shoot-name}.kubeconfig` secrets in each `Project` namespace.
+
+### Non-Goals
+
+- Generate `OpenID Connect` kubeconfigs
+
+## Proposal
+
+The `gardener-apiserver` would serve a new `shoots/adminkubeconfig` resource. It can only accept `CREATE` calls and accept `AdminKubeConfigRequest`. A `AdminKubeConfigRequest` would have the following structure:
+
+```yaml
+apiVersion: authentication.gardener.cloud/v1alpha1
+kind: AdminKubeConfigRequest
+spec:
+  expirationSeconds: 3600
+```
+
+Where `expirationSeconds` is the validity of the certificate in seconds. It this case it would be `1 hour`. The maximum validity of a `AdminKubeConfigRequest` is configured by `--admin-kubeconfig-max-token-expiration` flag in the `gardener-apiserver`.
+
+When such request is received, the API sever would find the `ShootState` associated with that cluster and generate a `kubeconfig`. The x509 client certificate would be signed by the `Shoot` cluster's CA and the user used in the subject's common name would be from the `User.Info` used to make the request.
+
+```yaml
+apiVersion: authentication.gardener.cloud/v1alpha1
+kind: AdminKubeConfigRequest
+spec:
+  expirationSeconds: 3600
+status:
+  expirationTimestamp: "2021-02-22T09:06:51Z"
+  kubeConfig: # this is normally base64-encoded, but decoded for the example
+    apiVersion: v1
+    clusters:
+    - cluster:
+        certificate-authority-data: LS0tLS1....
+        server: https://api.shoot-cluster
+      name: shoot-cluster-a
+    contexts:
+    - context:
+        cluster: shoot-cluster-a
+        user: shoot-cluster-a
+      name: shoot-cluster-a
+    current-context: shoot-cluster-a
+    kind: Config
+    preferences: {}
+    users:
+    - name: shoot-cluster-a
+      user:
+        client-certificate-data: LS0tLS1CRUd...
+        client-key-data: LS0tLS1CRUd...
+```
+
+New feature gate called `AdminKubeConfigRequest` would enable this API in the `gardener-apiserver`. The old `{shoot-name}.kubeconfig` would be kept, but deprecated and removed in the future.
```

Comment:
Why is the feature gate needed? If we have one for GAPI, we probably also need one for gardenlet to control whether the old `{shoot-name}.kubeconfig` secret is synced to the project namespace.

Reply:
This is a new API that and administrators could want to not enable this functionality. If there is a security issue - it's a quick way to disable it.

Reply:
OK, fair enough.